Mathlib/Order/PrimeSeparator.lean

```diff
@@ -0,0 +1,156 @@
+  rw [isPrime_iff_mem_or_mem]
+  intros a₁ a₂
+  contrapose!
+  intro ⟨ha₁, ha₂⟩
+
+  -- Consider the ideals J₁, J₂ generated by J ∪ {a₁} and J ∪ {a₂}, respectively.
+  let J₁ := J ⊔ principal a₁
+  let J₂ := J ⊔ principal a₂
+
+  -- For each i, Jᵢ is an ideal that contains aᵢ, and is not equal to J.
+
+  have a₁J₁ : a₁ ∈ J₁ := mem_of_subset_of_mem (le_sup_right : _ ≤ J ⊔ _) mem_principal_self
+  have a₂J₂ : a₂ ∈ J₂ := mem_of_subset_of_mem (le_sup_right : _ ≤ J ⊔ _) mem_principal_self
+  have J₁J : ↑J₁ ≠ Jset := by refine ne_of_mem_of_not_mem' a₁J₁ ha₁
+  have J₂J : ↑J₂ ≠ Jset := by refine ne_of_mem_of_not_mem' a₂J₂ ha₂
+
+  -- Therefore, since J is maximal, we must have Jᵢ ∉ S.
+  have J₁S : ↑J₁ ∉ S := fun h => J₁J (Jmax J₁ h (le_sup_left : J ≤ J₁))
+  have J₂S : ↑J₂ ∉ S := fun h => J₂J (Jmax J₂ h (le_sup_left : J ≤ J₂))
+
+  -- Since Jᵢ is an ideal that contains I, we have that Jᵢ is not disjoint from F.
+  have J₁F : ¬ (Disjoint (F : Set α) J₁) := by
+    intro hdis
+    apply J₁S
+    simp only [le_eq_subset, mem_setOf_eq, SetLike.coe_subset_coe, S]
+    exact ⟨J₁.isIdeal, le_trans IJ' le_sup_left, hdis⟩
+
+  have J₂F : ¬ (Disjoint (F : Set α) J₂) := by
+    intro hdis
+    apply J₂S
+    simp only [le_eq_subset, mem_setOf_eq, SetLike.coe_subset_coe, S]
+    exact ⟨J₂.isIdeal, le_trans IJ' le_sup_left, hdis⟩
+
+  -- Thus, pick cᵢ ∈ F ∩ Jᵢ.
+  let ⟨c₁, ⟨c₁F, c₁J₁⟩⟩ := Set.not_disjoint_iff.1 J₁F
+  let ⟨c₂, ⟨c₂F, c₂J₂⟩⟩ := Set.not_disjoint_iff.1 J₂F
+
+  -- Using the definition of Jᵢ, we can pick bᵢ ∈ J such that cᵢ ≤ bᵢ ⊔ aᵢ.
+  let ⟨b₁, ⟨b₁J, cba₁⟩⟩ := (mem_ideal_sup_principal a₁ c₁ J).1 c₁J₁
+  let ⟨b₂, ⟨b₂J, cba₂⟩⟩ := (mem_ideal_sup_principal a₂ c₂ J).1 c₂J₂
+
+  -- Since J is an ideal, we have b := b₁ ⊔ b₂ ∈ J.
+  let b := b₁ ⊔ b₂
+  have bJ : b ∈ J := sup_mem b₁J b₂J
+
+  have ineq : c₁ ⊓ c₂ ≤ b ⊔ (a₁ ⊓ a₂) :=
+  -- Now we compute: b ⊔ (a₁ ⊓ a₂) = (b ⊔ a₁) ⊓ (b ⊔ a₂) by distributivity
+  -- and b ⊔ aᵢ ≥ bᵢ ⊔ aᵢ by definition of b and monotonicity of ⊔.
+  -- Thus, b ⊔ (a₁ ⊓ a₂) ≥ (b₁ ⊔ a₁) ⊓ (b₂ ⊔ a₂) ≥ c₁ ⊓ c₂,
+  -- with the last inequality holding by monotonicity of inf.
+  calc
+    c₁ ⊓ c₂ ≤ (b₁ ⊔ a₁) ⊓ (b₂ ⊔ a₂) := inf_le_inf cba₁ cba₂
+    _       ≤ (b  ⊔ a₁) ⊓ (b  ⊔ a₂) := by
+      apply inf_le_inf <;> apply sup_le_sup_right; exact le_sup_left; exact le_sup_right
+    _       = b ⊔ (a₁ ⊓ a₂) := (sup_inf_left b a₁ a₂).symm
+
+  -- Note that c₁ ⊓ c₂ ∈ F, since c₁ and c₂ are both in F and F is a filter.
+  -- Since F is an upper set, it now follows that b ⊔ (a₁ ⊓ a₂) ∈ F.
+  have ba₁a₂F : b ⊔ (a₁ ⊓ a₂) ∈ F := PFilter.mem_of_le ineq (PFilter.inf_mem c₁F c₂F)
+
+  -- Now, if we would have a₁ ⊓ a₂ ∈ J, then, since J is an ideal and b ∈ J, we would also get
+  -- b ⊔ (a₁ ⊓ a₂) ∈ J. But this contradicts that J is disjoint from F.
+
+  intro ha₁a₂
+  have notdis : ¬ (Disjoint (F : Set α) J) := by
+    rw [Set.not_disjoint_iff]
+    use b ⊔ (a₁ ⊓ a₂)
+    exact ⟨ba₁a₂F, sup_mem bJ ha₁a₂⟩
+  exact notdis JF
```

Comment:
With `contrapose` you can go from `(hP : P) ⊢ ¬Q` to `(hQ : Q) ⊢ ¬ P` in one tactic.
```suggestion
  contrapose! JF with ha₁a₂
  rw [Set.not_disjoint_iff]
  use b ⊔ (a₁ ⊓ a₂)
  exact ⟨ba₁a₂F, sup_mem bJ ha₁a₂⟩
```

Reply:
Ah great, this is what I was looking for!